### DISCUSSION TITLE
Timeout warnings on Home Office services

### DISCUSSION THREAD

**Discussion by jamesmckechnie** [UKHomeOffice/design-system#383] [Show and tell]
# Timeout warnings at the Home Office

This post only includes key information. Please read [COP-timeouts-functionality.pdf](https://github.com/UKHomeOffice/design-system/files/8571287/COP-timeouts-functionality.pdf) for full design specifications and functionality details which are useful for when speaking to developers.

## Background
I've been working with delivery teams across the Home Office to design, research and refine a pattern for timeout warnings. The aim was to build on what already exists on the [Home Office design system](https://design.homeoffice.gov.uk/patterns/stop-a-service-timing-out) and in other departments, like [HMRC](https://design.tax.service.gov.uk/hmrc-design-patterns/service-timeout/).

We completed rounds of design and research on:
- Status (UVKI public-facing service, part of MBTP)
- Apply (UVKI public-facing service, part of MBTP)
- COP (Border Force internal-facing service, part of MBTP) 
- Modern Slavery Statement Registry (public-facing service, part go PPPT

## Design 
- https://www.figma.com/file/9zABI8oAJSF9MvMiQgVp2d/Session-timeout?node-id=0%3A1 
- [COP-timeouts-functionality.pdf](https://github.com/UKHomeOffice/design-system/files/8571287/COP-timeouts-functionality.pdf)
![Flows](https://user-images.githubusercontent.com/34271054/165509307-f68b41d1-1427-4b67-ba13-3aae0fb0ce27.png)
![Desktop-modal-styles](https://user-images.githubusercontent.com/34271054/165509354-388537a4-1ac8-443f-af44-8f15e3164725.png)
![Mobile-modal-styles](https://user-images.githubusercontent.com/34271054/165509365-615d5eca-8efa-4777-86f4-ee665f39eca9.png)

## How to implement on GOV.UK prototyping kit
This pattern uses a jQuery accessible modal from https://a11y.nicolas-hoffmann.net/modal/.

‼️ This code can be very buggy, it was written by a designer an is meant for prototyping and demoing high level functionality. It shouldn't be used in a live service. 

💡TIP: Unzip these files before using them.

1. Copy code from [modal-and-timer.html.zip](https://github.com/UKHomeOffice/design-system/files/8571547/modal-and-timer.html.zip) to the html page you want the modal to display. There is a timer included to allow you to trigger the modal to open automatically and redirect the user to a message page if they don't react in time. 

2. Copy code from [modal.css.zip](https://github.com/UKHomeOffice/design-system/files/8571438/modal.css.zip) to an existing or new custom CSS file. 

💡 TIP: Make sure to call on this .css file in the `<head>` of your html page or `..views/includes/head.html`. 

3. Add the whole [jquery-accessible-modal.js.zip](https://github.com/UKHomeOffice/design-system/files/8571483/jquery-accessible-modal.js.zip) file to your `../assets/javascripts/` folder.

💡 TIP: Make sure to call on this .js file in the `<scripts>` of your html page or `..views/includes/scripts.html`. 

4. Add separate html pages for when the user chooses to sign out or is timed out by the service. 

5. Route the user to these pages using the `data-modal-close-route="signed-out"` and `<a id="timeout-link" href="timeout"></a>` parts of the html code

💡 TIP: You can easily edit the content of the modal in the .html file, for changes to the design you can change the css or the jQuery in the .js file too though this can have unexpected results. 

💡 TIP: Remove the timers in the html if you don't need them, they are often what causes issues.

## Research
We carried out research on different services in Home Office to gather data on different types of service. We used the similar design in all rounds of research with slightly altered content for each.

A caveat to the research in all rounds is that it was hard to recreate the true to life conditions in a research session where a user would be inactive for a long period of time. Instead, we opted to trigger the modal after two minutes to gauge a natural reaction, then debriefed the user later to explain that this would not normally happen.

### Apply (a public-facing UKVI service)

- Apply carried out research with 6 users for their service which is signed-in and allows users to enter data.
- When the warning appeared, most users quickly and instinctively pressed the green button to make it disappear.
- Both the concept of a timeout, as well as this type of warning, appeared to be familiar to users.
- Some users noted that they expected the answers they had entered should be saved, even if they do not act in time for them to be timed out.

### Status (a public-facing UKVI service)

- Status carried out research with 6 users for their service which is signed-in and mostly does not allow users to enter data. All users were understood to be confident digital users.
- When the warning appeared, all users closed the modal quickly without reading all the content. All users closed the warning using the primary call-to-action.
- Users have high awareness of timeouts and timeout warnings
- Users will often dismiss / acknowledge the timeout quickly without reading the content
- Users were often unaware of the secondary option as they closed the warning so quickly.
- All users understood that clicking the primary call-to-action would extend their session.
- When asked about the 5-minute timer, most users understood that this is how much time they had left to act.
- Most users were able to explain what the warning meant, they could explain that they would be logged out after a period of inactivity and that this was a security feature to keep their data safe. Most users had seen a similar feature on other services they have used.
- No users opted to close the modal by clicking outside, using the ESC key or to sign out.
- We were not able to test the message pages for this round.

### COP (an internal Border Force service)

- We carried out research with 6 users for their service which is signed-in and allows users to enter data. It is also an internal service, used mainly by Border Force staff.
- All users selected ‘continue on this page’, with little notice taken around the content - like it was a natural reaction to a notification / pop up modal.
- Mix of understanding / observation that it would sign them out in 5 minutes. They clicked continue so they could go back to what they were doing.
- They seemed to think it came up because of inactivity, or that it should come up when there is inactivity to support security.
- Some expectation that timer would be reset / they wouldn’t be signed out after clicking continue.
- No real expectation that the time would change on the message, however difficult to tell with this particular test because this had to be asked directly. One user linked to computer restart timer, so may expect similar behaviour.
- Some mentioning they would need more time to complete a form, so they may associate the timeout with how long they must complete the form.
General understanding it wouldn't save the form [when being timed out] but mentions that save would be useful - lines up with previous research.

### Gaps in research

More research could be done into:
- what users expect to happen with their data when they chose to sign out?
- when signing back in, do users expect to be taken back to where they were?
- how effective is the countdown timer and does it cause an unnecessary anxiety?
- when user closes the modal with either ‘ESC’ or clicking outside, do they understand that their session has been renewed?
- does the content on the ‘message pages’ correctly convey what has happened and what the user should do next?

Research with key user groups, including:
- Screen reader users
- Screen zoom users
- Keyboard-only users
- Users with cognitive impairments which could cause slower reaction times
- Users with lower digital literacy

### Testing the built pattern

Test the feature using assistive technologies using the built design for most accurate findings. We should test across browser types, device types and screen sizes.

### Analytics and tracking

It would be useful to see how the pattern performs in the wild. Using Google Analytics or Matomo could let a team gain insight into how often timeout warnings appear and how users behave when they do.

Some ideas of things to track:
- when timeout modal appears
- how quickly it is closed
- methods of closing the modal (ESC, clicking away or primary CTA)
- user actions (continue, sign out or do nothing)
- how often users are timed out
- user behaviour after timing out or signing out – do they sign back in?

**Comment by daniel-ac-martin** [comment]
My initial impressions:

~~I think one of the questions with this, is how would it work with server-side rendering?~~ (Sorry, just seen that this is addressed in the PDF.)

Then again perhaps we just show an error on the next page load and invite them to log-in?

This seems more important when you are only allowing a short log-in session. (But should you do that in the first place?) For context, it might be good to know what the time-out is on the services in question.

Disclaimer: I'm a dev, not a designer!

**Comment by guermanc** [comment]
Seems that the page structure currently deviates significantly from the way existing patterns are implemented. Would it be worth aligning it more closely to the existing pattern pages?

In addition, this will need an accessibility section detailing some of the considerations. We could use the acceptance criteria from https://gist.github.com/hannalaakso/2641fc16d2158e60d551cd9da960b5da#acceptance-criteria which covers the main accessibility considerations.

I haven't had a chance to try out the proposed modal, but it sounds to me that it is reliant on JavaScript. What considerations have been made in regards to progressive enhancement? How will users who have JavaScript disabled be able to extend their session, and if this is not possible without adding undue complexity, what is the workaround. A potential solution is discussed in https://github.com/alphagov/govuk-design-system-backlog/issues/104#issuecomment-394620328 which could be included.

**Comment by guermanc** [comment]
I've had a look at the accessibility of the jQuery accessible modal at https://a11y.nicolas-hoffmann.net/modal/. There were a couple of accessibility issues, and a couple of general suggestions.

## Issues

- WCAG 1.4.10 Reflow - At higher magnification levels up to 400%, contents of the modal become cut off at the bottom with no way to access it, even with a mouse
- WCAG 4.1.1 Parsing - Some parsing issues, but these appear to not be directly linked to the modal. This will need to be tested again once the code has been implemented
- WCAG 4.1.2 Name, Role, Value - The "close" area around the modal is announced at the bottom of the content without a role. This facilitates the clicking outside the modal to close it, but interactivity not announced to screen reader users

## Best Practice

- When using JAWS 2019 on Edge 100, the "close" area around the outside of the modal needs to be selected twice to close the modal. When using TalkBack, selecting this area does nothing
- Default focus indicator for buttons to open and close the modal is difficult to see on most browsers. This will undoubtedly change with the use of custom CSS
- Modals does not respect Android 11 Display Font Size and Chrome 102 Text Scaling settings

**Comment by guermanc** [comment]
The W3C now has the [Alerts and Message Dialogs pattern](https://www.w3.org/WAI/ARIA/apg/patterns/alertdialog/) in the ARIA Authoring Practices Guide. This could be a better, more accessible, example to base our implementation on and is implemented in JavaScript.

**Comment by Sarraounia** [comment]
Met to discuss how to move this forward, with @jamesmckechnie @willrichards and @guermanc 

Latest idea is to split this work into two and create:
1. A modal component 
2. A new general timeout pattern

The responsible and careful use of modals needs to be clearly signposted in the content for the component. There are well-documented accessibility issues with modals but instances where teams still need to use them, so we should provide some guidance on best practice for consistency. Consider the use of an 'experimental' tag?

We will need the help of a developer to create the code for the component. @willrichards said COP will be creating one soon, but there is no fixed timeline for this. 

The new timeout pattern likely should have a separate discussion to gather evidence and examples from HO services.

**Comment by mviegasb** [comment]
Has COP built this component yet? Would be interested in reusing if so @willrichards
